### PR TITLE
Validation Overhaul

### PR DESCRIPTION
**Pull request by rehtropsmb**
* pausing on the frame when the ball enters the goal tape no longer invalidates an IL attempt (woo!)
* unified validation from ilbattle and ilmark (in validation.cpp)
* moved framesave/subtick logic out of timer.cpp as well (to validation.cpp)

I'm going to try to get a lot of playtesting for this, I'm pretty sure the IL Mark is working fine but I want to make sure there isn't any weird behavior with IL battles (especially with the buzzer beaters)

### CODE REVIEW

**Comment by ComplexPlane** [comment]
Hm, are you sure that all of the validation logic should be shared? What if people want to do IL battles with variants? "Leaderboard-valid IL of vanilla stage" vs. "Valid IL attempt w/ current variants" seem slightly different to me

File: src/mods/validate.cpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+                find_framesave(ball, out_stage_goal_idx, out_itemgroup_id, out_goal_flags);
+                s_entered_goal = result;
+            }
+            return result;
+        });
+}
+
+void tick() {
+    if (mkb::sub_mode != mkb::SMD_GAME_GOAL_INIT) {
+    }
```
Comment (ComplexPlane) [inline src/mods/validate.cpp]: Remove

File: src/systems/main.cpp [hunk 2]
```diff
@@ -182,6 +186,20 @@ void init() {
...
+
+                    bool ilmark_valid = validate::was_run_valid(false);
+                    bool battle_valid = validate::was_run_valid(true);
+
+                    if (ilmark_valid) {
+                        ilmark::set_valid();
+                    }
+                    if (battle_valid) {
+                        ilbattle::track_valid_attempt();
+                    }
```
Comment (ComplexPlane) [inline src/systems/main.cpp]: Can ilmark/ilbattle grab this themselves?

File: src/mods/validate.cpp [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+                mkb::sub_mode != mkb::SMD_GAME_GOAL_REPLAY_MAIN) {
+                s_framesave = static_cast<u32>((start.z / (start.z - end.z)) * 100);
+            }
+            return true;
+        }
+    }
+}
+
+// Copy paste from ghidra of did_ball_enter_goal with slight changes to calculate s_framesave
+void find_framesave(mkb::Ball* ball, int* out_stage_goal_idx, int* out_itemgroup_id,
```
Comment (ComplexPlane) [inline src/mods/validate.cpp:131]: Framesave calculation seems pretty orthogonal to IL validation and maybe ought to go in a different file eventually, but I'll let it slide for this PR - it makes sense to factor it out of timer.cpp at least

File: src/systems/main.cpp (hunk 2 above)
Comment (rehtropsmb) [reply src/systems/main.cpp]: Yup, moved it out of main and to their own respective functions

**Pull request merged**